transport/bdp_estimator.go

```diff
@@ -0,0 +1,47 @@
+package transport
+
+import (
+	"sync"
+)
+
+type bdpEstimator struct {
+	p                *ping
+	maxWindow        uint32
+	mu               sync.Mutex
+	estimate         uint32
+	accumulator      uint32 // Number of bytes received.
+	isSent           bool
+	send             func(*ping)    // Callback to send ping.
+	updateConnWindow func(n uint32) // Callback to update window size.
+}
+
+func (b *bdpEstimator) add(n uint32) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	if !b.isSent {
+		b.isSent = true
+		b.accumulator = 0
+		b.send(b.p)
+	}
+	b.accumulator += n
+}
+
+func (b *bdpEstimator) calculate(d [8]byte) {
```

Comment:
what algorithm is this trying to implement? `if estimate > b.estimate ` would seems to guarantee that one only ever increases, and never decreases.

What I would think you would want here is an exponential moving average; to also maintain a standard deviation in an online fashion, and then give variance weighted estimates by combining the two using Satterthwaite approximation so that you aren't assuming constant variance. Here are examples of what I mean.

https://github.com/glycerine/go-sliding-window/blob/master/rtt.go#L54
https://github.com/glycerine/go-sliding-window/blob/master/stddev.go#L80
https://github.com/glycerine/go-sliding-window/blob/master/sender.go#L677